core/state/database.go

```diff
@@ -253,36 +254,92 @@ func (tds *TrieDbState) Copy() *TrieDbState {
 	}
 
-	if !tds.noIntermediateCache {
-		cpy.t.SetUnloadFunc(cpy.putIntermediateCache)
-		cpy.tp.SetCreateNodeFunc(cpy.delIntermediateCache)
+	if debug.IsIntermediateTrieHash() {
+		cpy.t.SetUnloadNodeFunc(cpy.putIntermediateHash)
+		cpy.t.SetOnDeleteSubtreeFunc(tds.markSubtreeEmptyInIntermediateHash)
+		cpy.tp.SetCreateNodeFunc(cpy.delIntermediateHash)
 	}
 
 	return &cpy
 }
 
-func (tds *TrieDbState) putIntermediateCache(prefix []byte, nodeHash []byte) {
-	if len(prefix)%2 == 1 { // only put to bucket prefixes with even number of nibbles
+func (tds *TrieDbState) markSubtreeEmptyInIntermediateHash(prefix []byte) {
+	if len(prefix) != common.HashLength {
 		return
 	}
-	if len(prefix) == 0 {
+
+	db := tds.db
+	if hasDb, ok := tds.db.(ethdb.HasDb); ok {
+		db = hasDb.DB()
+	}
+
+	if err := db.Put(dbutils.IntermediateTrieHashBucket, prefix, []byte{}); err != nil {
+		log.Warn("could not put intermediate trie hash", "err", err)
+	}
+}
+
+func (tds *TrieDbState) putIntermediateHash(prefixAsNibbles []byte, nodeHash []byte) {
+	if len(prefixAsNibbles) == 0 {
 		return
 	}
-	if err := putIntermediateCache(tds.db, prefix, nodeHash); err != nil {
-		log.Warn("could not put intermediate trie cache", "err", err)
+
+	if len(prefixAsNibbles)%2 == 1 { // only put to bucket prefixes with even number of prefixAsNibbles
+		return
+	}
+
+	key := pool.GetBuffer(64)
+	defer pool.PutBuffer(key)
+
+	if err := trie.CompressNibbles(prefixAsNibbles, &key.B); err != nil {
+		log.Warn("could not CompressNibbles for intermediate trie hash", "err", err)
+		return
+	}
+
+	v := common.CopyBytes(nodeHash)
+
+	db := tds.db
+	if hasDb, ok := tds.db.(ethdb.HasDb); ok {
+		db = hasDb.DB()
+	} else {
+		fmt.Printf("what is it? %T\n", tds.db)
 	}
 
+	//	fmt.Printf("Put: %x %x\n", key.Bytes(), v)
+	if err := db.Put(dbutils.IntermediateTrieHashBucket, key.Bytes(), v); err != nil {
+		log.Warn("could not put intermediate trie hash", "err", err)
+	}
+	_, _ = db, v
 }
 
-func (tds *TrieDbState) delIntermediateCache(prefix []byte) {
-	if len(prefix)%2 == 1 { // only put to bucket prefixes with even number of nibbles
+func (tds *TrieDbState) delIntermediateHash(prefixAsNibbles []byte) {
+	if len(prefixAsNibbles) == 0 {
 		return
 	}
-	if len(prefix) == 0 {
+
+	if len(prefixAsNibbles)%2 == 1 { // only put to bucket prefixes with even number of nibbles
 		return
 	}
-	if err := delIntermediateCache(tds.db, prefix); err != nil {
-		log.Warn("could not delete intermediate trie cache", "err", err)
+
+	key := pool.GetBuffer(64)
+	defer pool.PutBuffer(key)
+
+	if err := trie.CompressNibbles(prefixAsNibbles, &key.B); err != nil {
+		log.Warn("could not CompressNibbles for intermediate trie hash", "err", err)
+		return
+	}
+
+	db := tds.db
+	if hasDb, ok := tds.db.(ethdb.HasDb); ok {
+		db = hasDb.DB()
+	} else {
+		fmt.Printf("what is it? %T\n", tds.db)
+	}
+
+	//fmt.Printf("Del: %x\n", key.Bytes())
+
+	if err := db.Delete(dbutils.IntermediateTrieHashBucket, key.Bytes()); err != nil {
```

Comment:
The same comment as in the `putIntermediateHash`, please do not use direct DB access, but simply copy the key